### PR TITLE
Support String to Enum conversion

### PR DESCRIPTION
**Pull request by bhaarathms**
### Description

*Support String to Enum conversion*

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs [hunk 1]
```diff
@@ -261,6 +261,13 @@ internal static void GetTypeFacets(IEdmTypeReference type, out int? precision, o
                     return true;
                 }
 
+                // Comparing an enum with a string is valid
+                if (left.IsEnum() && right.IsString())
+                {
+                    right = left;
+                    return true;
+                }
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs]: Any reason not to support the same comparison with the string on the left and enum on the right?

File: src/Microsoft.OData.Core/UriParser/Binders/BinaryOperatorBinder.cs [hunk 2]
```diff
@@ -72,6 +72,12 @@ internal static void PromoteOperandTypes(BinaryOperatorKind binaryOperatorKind,
                 throw new ODataException(ODataErrorStrings.MetadataBinder_IncompatibleOperandsError(leftTypeName, rightTypeName, binaryOperatorKind));
             }
 
+            // If comparing an enum with a string, we must convert the string into the enum equivalent
+            if (left.TypeReference.IsEnum() && right.TypeReference.IsString())
+            {
+                right = new ConstantNode(new ODataEnumValue(((ConstantNode)right).Value.ToString(), leftType.Definition.ToString()), ((ConstantNode)right).Value.ToString(), rightType);
+            }
+
```
Comment (mikepizzo) [inline src/Microsoft.OData.Core/UriParser/Binders/BinaryOperatorBinder.cs]: After PromoteOperandTypes sets the type of the string operand to the appropriate enum type, would it be better to put the logic to create the appropriate constant node within  MetadataBindingUtils.ConvertToTypeIfNeeded?

**Comment by mikepizzo** [comment]
Thanks for the contribution; eager to get this one integrated!  OData 4.01 adds support for implicit conversions of string to both enums and duration values; any interest in expanding this PR to also support string to duration conversions?

File: src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs (hunk 1 above)
Comment (bhaarathms) [reply src/Microsoft.OData.Core/UriParser/TypePromotionUtils.cs]: That looks like a valid case. Added logic for conversion when string is on left

File: src/Microsoft.OData.Core/UriParser/Binders/BinaryOperatorBinder.cs (hunk 2 above)
Comment (bhaarathms) [reply src/Microsoft.OData.Core/UriParser/Binders/BinaryOperatorBinder.cs]: Sure. Moved the conversion logic to ConvertToTypeIfNeeded. Looks cleaner this way.

**Comment by bhaarathms** [comment]
@mikepizzo Thank you for taking a look at the PR. 
I think I would have to do some research on duration conversions. Right now I have few other things on my schedule. I will check back in a week and try adding them if they are not added by then.

**Comment by AlanWong-MS** [comment]
@bhaarathms, heads up: we recently made committed some changes to the master branch that you'll have to pull. The build check will fail without the new changes.

**Comment by bhaarathms** [comment]
Looks like I messed something up during rebase. I will wait to see the build status. I will create a new pull request and close this one

**Comment by bhaarathms** [comment]
@mikepizzo created a new PR as I messed this one up. https://github.com/OData/odata.net/pull/980 is this new PR. Please take a look.

**Pull request closed without merging**